magic_items/services.py

```diff
@@ -0,0 +1,358 @@
+    def get(
+        cls,
+        item_id: int,
+    ) -> MagicItemDomain:
+        item_model = MagicItemModel.query.get(item_id)
+        return MagicItemDomain.from_model(item_model)
+
+    @classmethod
+    def get_filtered(
+        cls,
+        **filters: dict,
+    ) -> List[MagicItemDomain]:
+
+        edited_filter = {}
+
+        for filter in filters:
+            if filters[filter] is not "":
+                edited_filter.update({filter: filters[filter]})
+
+        model_magic_items = MagicItemModel.query.filter_by(**edited_filter).all()
+        domain_magic_items = []
+
+        for magic_item_model in model_magic_items:
+            magic_item_domain = MagicItemDomain.from_model(magic_item_model)
+            domain_magic_items.append(magic_item_domain)
+
+        return domain_magic_items
+
+    @classmethod
+    def update(cls, user_id: int, magic_item_id: int, form: ItemForm):
+
+        item_model = MagicItemModel.query.get_or_404(magic_item_id)
+
+        for k, v in form.data.items():
+            if k != CSRF_TOKEN and k != DESCRIPTION:
+                setattr(item_model, k, v)
+
+        item_model.description = (
+            f"{form.rarity.data}, {form.item_type.data}",
+            form.description.data,
+        )
+        item_model.last_updated = datetime.utcnow()
+
+        db.session.commit()
+
+        return MagicItemDomain.from_model(item_model)
+
+    @classmethod
+    def delete(cls, user_id: int, magic_item_id: int) -> bool:
+
+        item_model = MagicItemModel.query.get_or_404(magic_item_id)
+
+        if item_model.created_by is not user_id:
+            abort(403)
+
+        db.session.delete(item_model)
+        db.session.commit()
+        return True
+
+    @classmethod
+    def random(
+        cls,
+    ) -> MagicItemDomain:
+
+        item_model = (
+            MagicItemModel.query.options(load_only("id"))
+            .offset(
+                func.floor(
+                    func.random() * db.session.query(func.count(MagicItemModel.id))
+                )
+            )
+            .limit(1)
+            .all()[0]
+        )
+
+        return MagicItemDomain.from_model(item_model)
+
+
+class CollectionService:
+    @classmethod
+    def create(
+        cls,
+        user_id: int,
+        form: CollectionAddForm,
+    ) -> CollectionDomain:
+
+        data = {k: v for k, v in form.data.items() if k != CSRF_TOKEN}
+
+        collection_model = CollectionModel(
+            **data,
+            user_id=user_id,
+        )
+
+        db.session.add(collection_model)
+        db.session.commit()
+
+        return CollectionDomain.from_model(collection_model)
+
+    @classmethod
+    def update(
+        cls,
+        user_id: int,
+        collection_id: int,
+        form: CollectionAddForm,
+    ) -> CollectionDomain:
+
+        collection_model = CollectionModel.query.get_or_404(collection_id)
+
+        if collection_model.user_id is not user_id:
+            abort(403)
```

Comment:
aborting something like this is the route's responsibility. I'd raise an exception here and catch it in the route then abort(403) there